encryption/src/main/java/org/apache/solr/update/DirectUpdateHandler2.java

```diff
@@ -0,0 +1,1116 @@
+    numDocsPending.reset();
+    try {
+      super.close();
+    } catch (Exception e) {
+      throw new IOException("Error closing", e);
+    }
+  }
+
+  // IndexWriterCloser interface method - called from solrCoreState.decref(this)
+  @Override
+  public void closeWriter(IndexWriter writer) throws IOException {
+    log.trace("closeWriter({}): ulog={}", writer, ulog);
+
+    assert TestInjection.injectNonGracefullClose(core.getCoreContainer());
+
+    boolean clearRequestInfo = false;
+
+    SolrQueryRequest req = new LocalSolrQueryRequest(core, new ModifiableSolrParams());
+    SolrQueryResponse rsp = new SolrQueryResponse();
+//    if (SolrRequestInfo.getRequestInfo() == null) {
+//      clearRequestInfo = true;
+//      SolrRequestInfo.setRequestInfo(new SolrRequestInfo(req, rsp)); // important for debugging
+//    }
+    try {
+
+      if (TestInjection.injectSkipIndexWriterCommitOnClose(writer)) {
+        // if this TestInjection triggers, we do some simple rollback()
+        // (which closes the underlying IndexWriter) and then return immediately
+        log.warn("Skipping commit for IndexWriter.close() due to TestInjection");
+        if (writer != null) {
+          writer.rollback();
+        }
+        // we shouldn't close the transaction logs either, but leaving them open
+        // means we can't delete them on windows (needed for tests)
+        if (ulog != null) ulog.close(false);
+
+        return;
+      }
+
+      // do a commit before we quit?
+      boolean tryToCommit =
+          writer != null
+              && ulog != null
+              && ulog.hasUncommittedChanges()
+              && ulog.getState() == UpdateLog.State.ACTIVE;
+
+      // be tactical with this lock! closing the updatelog can deadlock when it tries to commit
+      solrCoreState.getCommitLock().lock();
+      try {
+        try {
+          if (log.isInfoEnabled()) {
+            log.info(
+                "Committing on IndexWriter.close() {}.",
+                (tryToCommit ? "" : " ... SKIPPED (unnecessary)"));
+          }
+          if (tryToCommit) {
+            CommitUpdateCommand cmd = new CommitUpdateCommand(req, false);
+            cmd.openSearcher = false;
+            cmd.waitSearcher = false;
+            cmd.softCommit = false;
+
+            // TODO: keep other commit callbacks from being called?
+            // this.commit(cmd); // too many test failures using this method... is it because of
+            // callbacks?
+
+            synchronized (solrCoreState.getUpdateLock()) {
+              ulog.preCommit(cmd);
+            }
+
+            // todo: refactor this shared code (or figure out why a real CommitUpdateCommand can't
+            // be used)
+            //TODO: change for encryption - start
+            if (shouldCommit(cmd, writer)) {
+              SolrIndexWriter.setCommitData(writer, cmd.getVersion(), cmd.commitData);
+              writer.commit();
+            }
+            //TODO: change for encryption - end
```

Comment:
This is the small change required for encryption. I'll do the change upstream.
Actually, from the comment line 959 above, ideally we should call commit(cmd) here instead of copying the code from the commit() method. But apparently many tests failed. So, given this code is a copy of commit(), we should also copy the shouldCommit() call part.